spec.html

```diff
@@ -648,28 +651,46 @@ <h1><ins>ExecuteAsyncModule ( _module_ )</ins></h1>
+      <emu-alg>
+        1. Assert: _module_.[[Status]] is ~evaluated~.
+        1. For each Module _m_ of _module_.[[AsyncParentModules]], do
+          1. If _execList_ does not contain _m_ and _m_.[[CycleRoot]].[[EvaluationError]] is ~empty~, then
+            1. Assert: _m_.[[EvaluationError]] is ~empty~.
+            1. Assert: _m_.[[AsyncEvaluating]] is *true*.
+            1. Assert: _m_.[[PendingAsyncDependencies]] > 0.
+            1. Set _m_.[[PendingAsyncDependencies]] to _m_.[[PendingAsyncDependencies]] - 1.
+            1. If _m_.[[PendingAsyncDependencies]] is equal to 0, then
+                1. Append _m_ to _execList_.
+                1. If _m_.[[Async]] is *false*, perform ! GatherAsyncParentCompletions(_m_, _execList_).
+        1. Return *undefined*.
+      </emu-alg>
+      <emu-note>
+        <p>When an async execution for a root _module_ is fulfilled, this function determines the list of modules which are able to synchronously execute together on this completion, populating them in _execList_.</p>
+      </emu-note>
+    </emu-clause>
+
     <emu-clause id="sec-async-module-execution-fulfilled" aoid="AsyncModuleExecutionFulfilled">
       <h1><ins>AsyncModuleExecutionFulfilled ( _module_ )</ins></h1>
       <emu-alg>
-        1. Assert: _module_.[[Status]] is ~evaluated~.
-        1. If _module_.[[AsyncEvaluating]] is *false*,
-          1. Assert: _module_.[[EvaluationError]] is not ~empty~.
-          1. Return *undefined*.
+        1. Assert: _module_.[[AsyncEvaluating]] is *true*.
         1. Assert: _module_.[[EvaluationError]] is ~empty~.
         1. Set _module_.[[AsyncEvaluating]] to *false*.
-        1. For each Module _m_ of _module_.[[AsyncParentModules]], do
-          1. Decrement _m_.[[PendingAsyncDependencies]] by 1.
-          1. If _m_.[[PendingAsyncDependencies]] is 0 and _m_.[[EvaluationError]] is ~empty~, then
-            1. Assert: _m_.[[AsyncEvaluating]] is *true*.
-            1. If _m_.[[CycleRoot]].[[EvaluationError]] is not ~empty~, return *undefined*.
-            1. If _m_.[[Async]] is *true*, then
-              1. Perform ! ExecuteAsyncModule(_m_).
-            1. Otherwise,
-              1. Let _result_ be _m_.ExecuteModule().
-              1. If _result_ is a normal completion,
-                1. Perform ! AsyncModuleExecutionFulfilled(_m_).
-              1. Otherwise,
-                1. Perform ! AsyncModuleExecutionRejected(_m_, _result_.[[Value]]).
+        1. Let _execList_ be a new empty List.
+        1. Perform ! GatherAsyncParentCompletions(_module_, _execList_).
+        1. Let _sortedExecList_ be a List of elements that are the elements of _execList_, in the order in which they had their [[AsyncEvaluating]] fields set to *true* in InnerModuleEvaluation.
+        1. Assert: All elements of _sortedExecList_ have their [[AsyncEvaluating]] field set to *true*, [[PendingAsyncDependencies]] field set to 0 and [[EvaluationError]] field set to *undefined*.
```

Comment:
There's a bug here: the top-level capability might never get resolved. First part of the fix is to insert a step here that does:

`1. Let _topLevelParent_ be _module_.`